# Detect financial fraud with decision tree in Python
07-detect-financial-fraud-w-dt-in-python

In [1]:
# Setup libraries
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [2]:
# Read data into dataframe
df = pd.read_csv('../00-Datasets/finance.csv')

C:\Users\jamie\AppData\Local\Temp\ipykernel_1964\4057374654.py:2: DtypeWarning: Columns (1,3,6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../00-Datasets/finance.csv')


In [3]:
df.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  24999.000000  2.499900e+04   2.499900e+04    2.499900e+04   
mean       6.311452  1.322446e+05   8.004936e+05    8.172910e+05   
std        2.368394  2.981803e+05   2.197915e+06    2.239935e+06   
min        1.000000  2.390000e+00   0.000000e+00    0.000000e+00   
25%        6.000000  5.771545e+03   0.000000e+00    0.000000e+00   
50%        7.000000  1.831492e+04   1.959100e+04    4.259590e+03   
75%        8.000000  1.549171e+05   1.382585e+05    1.408327e+05   
max        8.000000  1.000000e+07   2.240000e+07    2.250000e+07   

       oldbalanceDest  newbalanceDest       isFraud  isFlaggedFraud  
count    2.499900e+04    2.499900e+04  24999.000000         24999.0  
mean     8.497170e+05    1.205839e+06      0.003320             0.0  
std      2.523119e+06    3.148265e+06      0.057526             0.0  
min      0.000000e+00    0.000000e+00      0.000000             0.0  
25%      0.000000e+00    0.000000e+00      0.000000             0.0  
50%      0.000000e+00    0.000000e+00      0.000000             0.0  
75%      3.428501e+05    6.068785e+05      0.000000             0.0  
max      2.500000e+07    2.880000e+07      1.000000             0.0

In [4]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0   1.0   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1   1.0   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2   1.0  TRANSFER    181.00  C1305486145          181.0            0.00   
3   1.0  CASH_OUT    181.00   C840083671          181.0            0.00   
4   1.0   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0      0.0             0.0  
1  M2044282225             0.0             0.0      0.0             0.0  
2   C553264065             0.0             0.0      1.0             0.0  
3    C38997010         21182.0             0.0      1.0             0.0  
4  M1230701703             0.0             0.0      0.0             0.0

In [5]:
df.tail()

step type  amount nameOrig  oldbalanceOrg  newbalanceOrig nameDest  \
1048557   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048558   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048559   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048560   NaN  NaN     NaN      NaN            NaN             NaN      NaN   
1048561   NaN  NaN     NaN      NaN            NaN             NaN      NaN   

         oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
1048557             NaN             NaN      NaN             NaN  
1048558             NaN             NaN      NaN             NaN  
1048559             NaN             NaN      NaN             NaN  
1048560             NaN             NaN      NaN             NaN  
1048561             NaN             NaN      NaN             NaN

In [6]:
df.sample(6)

step      type    amount    nameOrig  oldbalanceOrg  newbalanceOrig  \
5494     5.0  CASH_OUT  89569.75  C719518227            0.0             0.0   
571642   NaN       NaN       NaN         NaN            NaN             NaN   
752746   NaN       NaN       NaN         NaN            NaN             NaN   
754578   NaN       NaN       NaN         NaN            NaN             NaN   
534243   NaN       NaN       NaN         NaN            NaN             NaN   
430029   NaN       NaN       NaN         NaN            NaN             NaN   

          nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
5494    C380242442        132427.3        19525.79      0.0             0.0  
571642         NaN             NaN             NaN      NaN             NaN  
752746         NaN             NaN             NaN      NaN             NaN  
754578         NaN             NaN             NaN      NaN             NaN  
534243         NaN             NaN             NaN      NaN             NaN  
430029         NaN             NaN             NaN      NaN             NaN

In [7]:
df.isna().sum()

step              1023563
type              1023563
amount            1023563
nameOrig          1023563
oldbalanceOrg     1023563
newbalanceOrig    1023563
nameDest          1023563
oldbalanceDest    1023563
newbalanceDest    1023563
isFraud           1023563
isFlaggedFraud    1023563
dtype: int64

In [8]:
df.dropna(inplace=True)

In [9]:
df.duplicated().sum()

0

In [10]:
df['isFlaggedFraud'].value_counts()

0.0    24999
Name: isFlaggedFraud, dtype: int64

In [11]:
df[df['isFlaggedFraud']!=0.0]

Empty DataFrame
Columns: [step, type, amount, nameOrig, oldbalanceOrg, newbalanceOrig, nameDest, oldbalanceDest, newbalanceDest, isFraud, isFlaggedFraud]
Index: []

In [23]:
# Though this isn't good quality data, 
# we'll continue with the exercise for 
# the purpose of practicing the algorithm. 

In [12]:
# Explore some features
df['type'].value_counts()

PAYMENT     12482
CASH_OUT     5091
CASH_IN      4505
TRANSFER     2425
DEBIT         496
Name: type, dtype: int64

In [13]:
# Explore tabular summary
df.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  24999.000000  2.499900e+04   2.499900e+04    2.499900e+04   
mean       6.311452  1.322446e+05   8.004936e+05    8.172910e+05   
std        2.368394  2.981803e+05   2.197915e+06    2.239935e+06   
min        1.000000  2.390000e+00   0.000000e+00    0.000000e+00   
25%        6.000000  5.771545e+03   0.000000e+00    0.000000e+00   
50%        7.000000  1.831492e+04   1.959100e+04    4.259590e+03   
75%        8.000000  1.549171e+05   1.382585e+05    1.408327e+05   
max        8.000000  1.000000e+07   2.240000e+07    2.250000e+07   

       oldbalanceDest  newbalanceDest       isFraud  isFlaggedFraud  
count    2.499900e+04    2.499900e+04  24999.000000         24999.0  
mean     8.497170e+05    1.205839e+06      0.003320             0.0  
std      2.523119e+06    3.148265e+06      0.057526             0.0  
min      0.000000e+00    0.000000e+00      0.000000             0.0  
25%      0.000000e+00    0.000000e+00      0.000000             0.0  
50%      0.000000e+00    0.000000e+00      0.000000             0.0  
75%      3.428501e+05    6.068785e+05      0.000000             0.0  
max      2.500000e+07    2.880000e+07      1.000000             0.0

In [15]:
# Prepare the data
le = LabelEncoder()
df['type'] = le.fit_transform(df['type'])

In [17]:
# Split data into 80% train and 20% set
X = df.drop(columns=['isFlaggedFraud', 'nameDest', 'nameOrig'], axis=1)
y = df['isFlaggedFraud']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [19]:
# Build model
model = DecisionTreeClassifier()
model.fit(X_train, y_train)

DecisionTreeClassifier()

In [21]:
# Feed test data into model
y_predict = model.predict(X_test)

actual_vs_predict = pd.DataFrame({'Actual': y_test,
                                 'Prediction': y_predict})
actual_vs_predict.sample(12)

Actual  Prediction
2549      0.0         0.0
7926      0.0         0.0
8152      0.0         0.0
4100      0.0         0.0
17640     0.0         0.0
8705      0.0         0.0
22315     0.0         0.0
24375     0.0         0.0
13000     0.0         0.0
8039      0.0         0.0
11800     0.0         0.0
14740     0.0         0.0

In [22]:
# Evaluate model
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00      5000

    accuracy                           1.00      5000
   macro avg       1.00      1.00      1.00      5000
weighted avg       1.00      1.00      1.00      5000



## Notes
* We wanted to detect fraudulent activities. 
* Our dataset had 1048562 records and 11 features, 8 of which are of data type float and 3 is of type object/string.
* There were a lot of missing values. All 11 columns of a row would be empty. After dropping these records, we are left with 24999 samples.
* A quick inspection showed that there are no duplicate records.
* A closer look at the label showed total imbalance. All records being labeled as not fraud.
* Though this isn't good quality data, we'll continue with the exercise for the purpose of practicing the algorithm. 
* The dataset was split into 80% train and 20% test sets. No validation set was created or used. No hyperparameter tuning occurred.
* We built, trained, and tested a decision tree classifier.
* We evaluated our model using ground truth and several metrics including precision, recall, and f1-score.